lib/writer/block-write-stream.js

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2017 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+const stream = require('readable-stream')
+const fs = require('fs')
+const debug = require('debug')('block-write-stream')
+
+const CHUNK_SIZE = 64 * 1024
+
+/**
+ * @summary BlockWriteStream
+ * @class
+ */
+class BlockWriteStream extends stream.Writable {
+  /**
+   * @summary BlockReadStream constructor
+   * @param {Object} [options] - options
+   * @param {Number} [options.fd] - file descriptor
+   * @param {String} [options.path] - file path
+   * @param {String} [options.flags] - file open flags
+   * @param {Number} [options.mode] - file mode
+   * @param {Boolean} [options.autoClose] - automatically close the stream on end
+   * @example
+   * new BlockWriteStream(options)
+   */
+  constructor (options) {
+    options = Object.assign({}, BlockWriteStream.defaults, options)
+    options.objectMode = true
+
+    debug('block-write-stream %j', options)
+
+    super(options)
+
+    this._writableState.highWaterMark = 1
+
+    this.fs = options.fs
+    this.fd = options.fd
+    this.path = options.path
+    this.flags = options.flags
+    this.mode = options.mode
+    this.autoClose = options.autoClose
+
+    this.position = 0
+    this.bytesRead = 0
+    this.blocksRead = 0
+    this.bytesWritten = 0
+    this.blocksWritten = 0
+
+    this.closed = false
+    this.destroyed = false
+
+    this.once('finish', function () {
+      if (this.autoClose) {
+        this.close()
+      }
+    })
+
+    this._flushing = false
+    this._firstBlocks = []
+
+    this.open()
+  }
+
+  /**
+   * @summary Internal write handler
+   * @private
+   * @param {Buffer} chunk - chunk buffer
+   * @param {String} encoding - chunk encoding
+   * @param {Function} next - callback(error, value)
+   * @example
+   * // Not to be called directly
+   */
+  _write (chunk, encoding, next) {
+    debug('_write', chunk.length, chunk.position, chunk.address)
+
+    // Wait for file handle to be open
+    if (this.fd == null) {
+      this.once('open', () => {
+        this._write(chunk, encoding, next)
+      })
+      return
+    }
+
+    this.bytesRead += chunk.length
+    this.blocksRead += 1
+
+    if (chunk.position == null) {
+      chunk.position = this.position
+    }
+
+    if (!this._flushing && (chunk.position < CHUNK_SIZE)) {
+      this._firstBlocks.push(chunk)
```

Comment:
It does only get used once – `_firstBlocks` is what contains the first blocks (which we need to write *last*, because Windows) of the stream, which are then flushed before end; this also obsoletes `bytesToZeroOutFromTheBeginning`

Reply:
Ah, I misunderstood what the code was doing then :-/  So is this a slight change in behaviour then, where previously we'd only write the first 512 bytes after everything else, and now we write the first 64KB after everything else?

I couldn't remember what `bytesToZeroOutFromTheBeginning` was about, so I did a bit of searching and it appears to be a bmap-thing rather than a Windows-thing #673